### PR TITLE
SNOW-979288: Add explicit DbType Parameter assignment

### PR DESCRIPTION
**Pull request by sfc-gh-ext-simba-lf**
### Description
Add explicit DbType Parameter assignment

### Checklist
- [ ] Code compiles correctly
- [ ] Code is formatted according to [Coding Conventions](../CodingConventions.md)
- [ ] Created tests which fail without the change (if possible)
- [ ] All tests passing (`dotnet test`)
- [ ] Extended the README / documentation, if necessary
- [ ] Provide JIRA issue id (if possible) or GitHub issue id in PR name

### CODE REVIEW

File: Snowflake.Data/Client/SnowflakeDbParameter.cs [hunk 1]
```diff
@@ -34,7 +36,28 @@ public SnowflakeDbParameter(int ParameterIndex, SFDataType SFDataType)
             this.SFDataType = SFDataType;
         }
 
-        public override DbType DbType { get; set; }
+        public override DbType DbType
+        {
+            get
+            {
+                if (_dbType != DbType.AnsiString || Value == null || Value is DBNull)
```
Comment (sfc-gh-mhofman) [inline Snowflake.Data/Client/SnowflakeDbParameter.cs:43]: what happens when _dbType and Value is null?

File: Snowflake.Data.Tests/UnitTests/SFDbParameterTest.cs [hunk 2]
```diff
@@ -125,5 +127,75 @@ public void TestDbParameterResetDbType([Values] SFDataType expectedSFDataType)
             _parameter.ResetDbType();
             Assert.AreEqual(SFDataType.None, _parameter.SFDataType);
         }
+
+        [Test]
+        public void TestDbTypeExplicitAssignment([Values] DbType expectedDbType)
```
Comment (sfc-gh-mhofman) [inline Snowflake.Data.Tests/UnitTests/SFDbParameterTest.cs:132]: what if DbType is not provided?
I would suggest to have a special edge case tests for:
- null value and (just one) non null DbType
- non null value and null DbType
- both - DbType and Value nulls

File: Snowflake.Data/Client/SnowflakeDbParameter.cs (hunk 1 above)
Comment (sfc-gh-ext-simba-lf) [reply Snowflake.Data/Client/SnowflakeDbParameter.cs]: I tried to test this but only `Value` can be set to null. The `_dbType` can't be set to null because of this error: `Cannot convert null to 'type' because it is a non-nullable value type`

It looks like `_dbType` will always default to `DbType.AnsiString` and the code would return the default `DbType`

File: Snowflake.Data.Tests/UnitTests/SFDbParameterTest.cs (hunk 2 above)
Comment (sfc-gh-ext-simba-lf) [reply Snowflake.Data.Tests/UnitTests/SFDbParameterTest.cs:132]: I added a test case for the first item on the list. And I also added a test case for null value and default DbType

But I'm not able to add a test case for the other two because `DbType` is a non-nullable value type. It always defaults to `DbType.AnsiString`

Comment (sfc-gh-mhofman) [reply Snowflake.Data.Tests/UnitTests/SFDbParameterTest.cs:132]: Thanks @sfc-gh-ext-simba-lf !

**Review by sfc-gh-mhofman — APPROVED** [review]
